README.md
# TractSeg

![Alt text](resources/Pipeline_img_v2.png)

Tool for fast and accurate white matter bundle segmentation from Diffusion MRI. It can create
bundle segmentations, segmentations of the endregions of bundles and Tract Orientation Maps (TOMs). Moreover, it can
do tracking on the TOMs creating bundle-specific tractogram and do Tractometry analysis on those.

The tool works very well for data similar to the Human Connectome Project. For other MRI datasets it works well for all
bundles except for the Commissure Anterior (CA) and the Fornix (FX) which are [incomplete sometimes](#small-bundles-like-the-ca-and-fx-are-incomplete).

TractSeg is the code for the following papers. Please cite the papers if you use it.

* Tract Segmentation:
[TractSeg - Fast and accurate white matter tract segmentation](https://doi.org/10.1016/j.neuroimage.2018.07.070) ([free arxiv version](https://arxiv.org/abs/1805.07103))
[NeuroImage 2018]
* Tract Orientation Mapping (TOM):
[Tract orientation mapping for bundle-specific tractography](https://arxiv.org/abs/1806.05580)
[MICCAI 2018]
* Tracking on TOMs:  
[Combined tract segmentation and orientation mapping for bundle-specific tractography](https://www.sciencedirect.com/science/article/pii/S136184151930101X)
[Medical Image Analysis 2019]
* Tractometry:  
[Multiparametric mapping of white matter microstructure in catatonia](https://www.nature.com/articles/s41386-020-0691-2) ([free preprint](resources/Wasserthal2020_Multiparametric_mapping_of_white_matter.pdf))
[Nature Neuropsychopharmacology 2020]

[![Brainlife](https://brainlife.io/api/warehouse/app/5b82d7f4e2f4f800275e020f/badge)](https://brainlife.io/app/5b82d7f4e2f4f800275e020f#)

## Table of contents

* [Install](#install)
* [How to use](#how-to-use)
* [FAQ](#faq)
* [Train your own model](#train-your-own-model)

## Install

TractSeg only runs on Linux and OSX. It works with Python >= 3.6.

#### Install Prerequisites

* [Pytorch](http://pytorch.org/)
* [Mrtrix 3](http://mrtrix.readthedocs.io/en/latest/installation/linux_install.html) (>= 3.0 RC3)
* [FSL](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FslInstallation) (if you already have a brain mask and do not use the
option `--preprocess` this is not needed)
* xvfb (`apt-get install xvfb`) (only needed if you use `plot_tractometry_results`)

#### Install TractSeg

Latest stable version:

```
pip install TractSeg
```

> NOTE: See [CHANGELOG](CHANGELOG.md) for (breaking) changes of each version  

If you want to use Cython for `Tracking` (speedup 2x) then install via

```
pip install https://github.com/MIC-DKFZ/TractSeg/archive/cython_tracking.zip
```

#### Docker

You can also directly use TractSeg via Docker (contains all prerequisites).

```
sudo docker run -v /absolute/path/to/my/data/directory:/data \
-t wasserth/tractseg_container:master TractSeg -i /data/my_diffusion_file.nii.gz -o /data --raw_diffusion_input
```

On OSX you might have to increase the Docker memory limit from the default of 2GB to something
like 7GB.

## How to use

#### Simple example

To segment the bundles on a Diffusion Nifti image run the following command. (Diffusion.bvals and Diffusion.bvecs have to be in the same directory
as the input image.)
(You can use the example image provided in this repository under `examples`.)  

```
TractSeg -i Diffusion.nii.gz --raw_diffusion_input
```

This will create a folder `tractseg_ouput` inside of the same directory as your input file with one binary segmentation nifti image
for each bundle.

> NOTE: If results look bad you probably have to align your images to have the same orientation as MNI space. Moreover it should have isotropic spacing. See [here](#aligning-image-to-mni-space) for more details.

#### Custom input and output path

```
TractSeg -i my/path/my_diffusion_image.nii.gz
         -o my/output/directory
         --bvals my/other/path/my.bvals
         --bvecs yet/another/path/my.bvec
         --raw_diffusion_input
```

Use `--help` to see all options.

#### Use existing peaks

To avoid generating the MRtrix CSD peaks every time you `run TractSeg you can also provide them directly by skipping the
option`--raw_diffusion_input`.

```
TractSeg -i my/path/my_mrtrix_csd_peaks.nii.gz
```

#### Segment bundle start and end regions

Get segmentations of the regions were the bundles start and end (helpful for filtering fibers that do not run
from start until end).

```
TractSeg -i peaks.nii.gz --output_type endings_segmentation
```

#### Create Tract Orientation Maps (TOMs)

For each bundle create a Tract Orientation Map ([Wasserthal et al., Tract orientation mapping for bundle-specific tractography](https://arxiv.org/abs/1806.05580)).
This gives you one peak per voxel telling you the main orientation of the respective bundle at this voxel.
Can be used for bundle-specific tracking later on.

```
TractSeg -i peaks.nii.gz --output_type TOM
```

> NOTE: `--output_type tract_segmentation` and `endings_segmentation` has to be run first. Same input and output
directories have to be used for all commands.

Peaks and streamlines can be visualized using for example [MITK Diffusion](http://mitk.org/wiki/DiffusionImaging#Downloads).
> NOTE: Peaks have to be flipped along the z-axis to be displayed correctly in MITK.  

#### Create bundle-specific tractograms

Tracks on TOMs and only keeps fibers not leaving the bundle mask and starting and ending in the endpoint regions.

```
TractSeg -i peaks.nii.gz --output_type tract_segmentation
TractSeg -i peaks.nii.gz --output_type endings_segmentation
TractSeg -i peaks.nii.gz --output_type TOM 
Tracking -i peaks.nii.gz
```

#### Use bedpostX peaks instead of CSD peaks

TractSeg also works with bedpostX as input. You have to pass `dyads1.nii.gz` as input and TractSeg will automatically
 find all the other necessary bedpostX files (`dyads2_thr0.05.nii.gz` & `dyads3_thr0.05.nii.gz`. `mean_f1-3samples`
 will be used for scaling the peaks.). This only works if you did not change the default bedpostX file naming.

```
TractSeg -i dyads1.nii.gz
```

#### Show uncertainty map

Create map showing where the method is uncertain about its segmentation (uses monte carlo dropout: <https://arxiv.org/abs/1506.02142>)

```
TractSeg -i peaks.nii.gz --uncertainty
```

#### Perform Tractometry

See [Documentation of Tractometry](resources/Tractometry_documentation.md).

#### Extended Tutorial

[Best pratices for standard usecases](resources/Tutorial.md).  
(You can also check out the [tutorial from OHBM 2019](https://colab.research.google.com/github/brainhack101/IntroDL/blob/master/notebooks/2019/Wasserthal/TractSegTutorial.ipynb)
[OUTDATED].)

#### Track subset of bundles

You can specify to only track a subset of bundles.

```
Tracking -i peaks.nii.gz --bundles CST_right,CA,IFO_right
```

#### Bundle names

The following list shows the index of each extracted bundle in the output file (if using `--single_output_file`).

```
0: AF_left         (Arcuate fascicle)
1: AF_right
2: ATR_left        (Anterior Thalamic Radiation)
3: ATR_right
4: CA              (Commissure Anterior)
5: CC_1            (Rostrum)
6: CC_2            (Genu)
7: CC_3            (Rostral body (Premotor))
8: CC_4            (Anterior midbody (Primary Motor))
9: CC_5            (Posterior midbody (Primary Somatosensory))
10: CC_6           (Isthmus)
11: CC_7           (Splenium)
12: CG_left        (Cingulum left)
13: CG_right   
14: CST_left       (Corticospinal tract)
15: CST_right 
16: MLF_left       (Middle longitudinal fascicle)
17: MLF_right
18: FPT_left       (Fronto-pontine tract)
19: FPT_right 
20: FX_left        (Fornix)
21: FX_right
22: ICP_left       (Inferior cerebellar peduncle)
23: ICP_right 
24: IFO_left       (Inferior occipito-frontal fascicle) 
25: IFO_right
26: ILF_left       (Inferior longitudinal fascicle) 
27: ILF_right 
28: MCP            (Middle cerebellar peduncle)
29: OR_left        (Optic radiation) 
30: OR_right
31: POPT_left      (Parieto‐occipital pontine)
32: POPT_right 
33: SCP_left       (Superior cerebellar peduncle)
34: SCP_right 
35: SLF_I_left     (Superior longitudinal fascicle I)
36: SLF_I_right 
37: SLF_II_left    (Superior longitudinal fascicle II)
38: SLF_II_right
39: SLF_III_left   (Superior longitudinal fascicle III)
40: SLF_III_right 
41: STR_left       (Superior Thalamic Radiation)
42: STR_right 
43: UF_left        (Uncinate fascicle) 
44: UF_right 
45: CC             (Corpus Callosum - all)
46: T_PREF_left    (Thalamo-prefrontal)
47: T_PREF_right 
48: T_PREM_left    (Thalamo-premotor)
49: T_PREM_right 
50: T_PREC_left    (Thalamo-precentral)
51: T_PREC_right 
52: T_POSTC_left   (Thalamo-postcentral)
53: T_POSTC_right 
54: T_PAR_left     (Thalamo-parietal)
55: T_PAR_right 
56: T_OCC_left     (Thalamo-occipital)
57: T_OCC_right 
58: ST_FO_left     (Striato-fronto-orbital)
59: ST_FO_right 
60: ST_PREF_left   (Striato-prefrontal)
61: ST_PREF_right 
62: ST_PREM_left   (Striato-premotor)
63: ST_PREM_right 
64: ST_PREC_left   (Striato-precentral)
65: ST_PREC_right 
66: ST_POSTC_left  (Striato-postcentral)
67: ST_POSTC_right
68: ST_PAR_left    (Striato-parietal)
69: ST_PAR_right 
70: ST_OCC_left    (Striato-occipital)
71: ST_OCC_right
```

#### Use different tract definitions

TractSeg was also trained on the bundles provided by [XTRACT](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/XTRACT).
These bundles are slightly differently defined. They tend to be more specific but therefore also a bit
less complete. Depending on your application this might be of interest for you. Using TractSeg instead of XTRACT
has the advantage of being at least 10x faster. Please cite XTRACT if you use this.
> NOTE: This is only supported for output type `tract_segmentation` and `dm_regression`.

```
TractSeg -i peaks.nii.gz --tract_definition xtract
```

#### Use python interface

```python
import nibabel as nib
import numpy as np
from tractseg.python_api import run_tractseg
peaks = nib.load("tests/reference_files/peaks.nii.gz").get_fdata()
segmentation = run_tractseg(peaks)
```

#### Different tracking types

You can use different types of tracking:

* "Probabilistic" tracking on TOM peaks [**default**].
`Tracking -i peaks.nii.gz`  
Probabilistic means that at each step a small random factor will be added to the direction given by the TOM peaks.
If not doing this on low resolution data it sometimes gets difficult finding fibers running from start to end and
covering the whole bundle.

* Probabilistic tracking on original FODs.
`Tracking -i WM_FODs.nii.gz --track_FODs iFOD2`  
Is calling Mrtrix iFOD2 tracking internally. Does not use TOM peaks but the original FODs. The results will get
filtered by the bundle mask and have to start and end in the endings masks.

#### Tracking formats

You can use the option `--tracking_format` to define the file format of the streamline files.
`tck` is the most stable tracking format and recommended. `trk` might get displayed differently in different viewers.  
> NOTE: When calling `Tractometry` and `plot_tractometry_results` you have to set the same tracking format as was used in `Tracking`.

#### Aligning image to MNI space

For best results the input image must have the same "orientation" as the Human Connectome Project data (MNI space)
(LEFT must be on the same side as LEFT of the HCP data) and have isotropic spacing.
If the image orientation and the gradient orientation of your data is the same as in `examples/Diffusion.nii.gz`
you are fine. Otherwise you should use `fslreorient2std` or rigidly register your image to MNI space (the brains
do not have to be perfectly aligned but must have the same LEFT/RIGHT orientation).
You can use the following FSL commands to rigidly register your image to MNI space (uses
the FA to calculate the transformation as this is more stable):

```shell
calc_FA -i Diffusion.nii.gz -o FA.nii.gz --bvals Diffusion.bvals --bvecs Diffusion.bvecs \
--brain_mask nodif_brain_mask.nii.gz

flirt -ref tractseg/tractseg/resources/MNI_FA_template.nii.gz -in FA.nii.gz \
-out FA_MNI.nii.gz -omat FA_2_MNI.mat -dof 6 -cost mutualinfo -searchcost mutualinfo

flirt -ref tractseg/tractseg/resources/MNI_FA_template.nii.gz -in Diffusion.nii.gz \
-out Diffusion_MNI.nii.gz -applyxfm -init FA_2_MNI.mat -dof 6
cp Diffusion.bvals Diffusion_MNI.bvals
rotate_bvecs -i Diffusion.bvecs -t FA_2_MNI.mat -o Diffusion_MNI.bvecs
```

To enforce isotropic spacing you can replace `-applyxfm` by `-applyisoxfm <your_spacing>`.

To move the results back to subject space you can use the following commands:

```shell
convert_xfm -omat MNI_2_FA.mat -inverse FA_2_MNI.mat  # invert transformation

flirt -ref FA.nii.gz -in my_bundle.nii.gz -out my_bundle_subject_space.nii.gz \
-applyxfm -init MNI_2_FA.mat -dof 6 -interp spline  # for TOM maps you have to use the command vecreg

fslmaths my_bundle_subject_space.nii.gz -thr 0.5 -bin my_bundle_subject_space.nii.gz  # float to binary 
```

The option `--preprocess` will automatically rigidly register the input image to MNI space, run TractSeg and then
convert the output back to subject space. This only works for `tract_segmentation` and `endings_segmentation`. For `TOMs` and trackings you have to register you data manually to MNI space:

```shell
# in first step --raw_diffusion_input has to be used together with --preprocess
TractSeg -i Diffusion.nii.gz -o tractseg_output --output_type tract_segmentation --raw_diffusion_input --preprocess
# -o has to be set in all following step
TractSeg -i tractseg_output/peaks.nii.gz -o tractseg_output --output_type endings_segmentation --preprocess
```

> NOTE: `--preprocess` does not work if you are using the option `--csd_type csd_msmt_5tt`,
because the T1 image will not automatically be registered to MNI space

If you are not familiar with preprocessing of DWI images (e.g. correcting for artifacts) you can have a look at
[this](resources/utility_scripts/dwi_preprocessing.sh) preprocessing script.

## FAQ

#### My output segmentation does not look like any bundle at all

Make sure your input image is in MNI space. Even if the input image is in MNI space the Mrtrix peaks might still be flipped.
In those cases you should view the peaks in `mrview` and make sure they have the proper
orientation. Otherwise you might have to flip the sign along the x, y or z axis using the following command:

```
flip_peaks -i my_peaks.nii.gz -o my_peaks_flip_y.nii.gz -a y
```

#### My data has anisotropic spacing. Does it still work?

If the spacing is only slightly anisotropic (e.g. 1.9mm x 1.9mm x 2mm) it still works fine. If the spacing is heavily
anisotropic (e.g. 1mm x 1mm x 2mm) results will be a lot better if you resample your image to isotropic spacing first.

#### Small bundles like the CA and FX are incomplete

You can use the following options to improve your results:  
`--super_resolution` The input image is upsampled to 1.25mm resolution (the resolution TractSeg was trained on) and
finally downsampled back to the original resolution. Using `--super_resolution` will output the image at 1.25mm.
Especially if image resolution is low parts of the CA can get lost during downsampling.

#### Can I save the pretrained weights at a different location?


CHANGELOG.md
# Changelog

## Master

* change sklearn to scikit-learn in setup.py

## Release 2.9

* bugfix: make download of weights from zenodo work again
* bugfix: change sklearn package name to scikit-learn in setup.py

## Release 2.8

* bugfix: reorient streamline in plot_tractometry_results was not in voxel space
* bugfix: use correct interpolation with flirt on binary masks
* xvfb for plotting

## Release 2.7

* Bugfix when tracking in non-MNI space
* **Interface change**: make `tck` the default tracking format since `trackvis` api is no longer supported by nibabel
* use spline interpolation instead of trilinear interpolation in flirt for option `--preprocess`
* improve weights download

## Release 2.6

* Update documentation
* fixed bug with newer fsl bet (thanks to @CarolynMcNabb)
* fix bug with import of `compress_streamline()` from dipy

## Release 2.5

* Move cython tracking to `cython_tracking` branch. It is not part of the main branch and pip package anymore because building the cython binaries for all possible operating system is a lot of pain. If people want to use it they can install from the `cython_tracking` branch directly which will automatically compile the cython code on their sytem.

## Release 2.4

* Update documentation
* Bugfix when tracking in non-MNI space
* Use Cython for Tracking. This will speed up tracking by roughly 2x and only require 1 CPU core instead of all CPU cores. Thanks to @StavrAspa and @eliaskoromilas for this contribution.

## Release 2.3

* Update documentation
* Update 3d tract visualization in `plot_tractometry_results` to use fury
* Updated pytorch to 1.8.1

## Release 2.2

* 3D plot of streamlines with coloring according to tractometry FA
* Add pretrained weights for XTRACT tract definitions
* Set range of y-axis in `plot_tractometry_results` (thanks to @elder-mama)
* Make `--preprocess` also work for endings_segmentation and TOM
* Simplified dockerfile
* Minor improvements

## Release 2.1.1

* Available on pypi
* Manually specify which bundles to track
* Minor improvements

## Release 2.1

* **Interface change**: The option `--bundle_specific_threshold` was removed. TractSeg checks itself now if CA or FX
are incomplete and then applies a lower threshold.
* **Interface change**: Postprocessing is activated by default now. If you want to deactivate is use `--no_postprocess`.
* minor improvements & Bugfixes
* FP16 training (increased training speed)
* Tractometry more testing and bugfix
* Tractometry now uses a far more advanced option to sample e.g. the FA along the tracts.
* Statistical analysis for tractometry data
* Python 2 not actively supported anymore (because dipy 1.0.0 does not support python 2 anymore)
* Add rotation (now peaks are also properly rotated) to data augmentation.
* Applies signs of affine to data if array not oriented like MNI data (needed to properly work with `fslreorient2std`)
* '--preprocess' will move output back to subject space.
* Updated weights for tract segmentation, endings segmentation and density regression (slightly increased
performance; now also trained with rotation during data augmentation; TOM not trained with rotation yet)
* Pretrained model also works with bedpostX peaks (instead of CSD peaks) (segmentation accuracy is the same)

## Release 2.0

* Increase training speed roughly by factor of 2 by using pin_memory and non_blocking for pytorch and by
cropping all non-brain area from the input images (requires preprocessing of the training data using
`tractseg/data/preprocessing.py`).
* Works with newer version of batchgenerators (Note: DataAugmentation slightly changed)
* Support bedpostX input
* Support aPTX tract definitions (but no pretrained model yet)
* Refactor `--preview`. Works without vtk now.
* Add plateau LR schedule to training
* Add API for mrtrix FACT tracking on TOMs
* Fix bug in rotation of bvecs when using option `--preprocess`.
* minor improvements
* Update TOM model and pretrained weights (Only angle in loss instead of angle and length. Gives slightly better
peak orientations.). Improved peak orientations allows for slightly less sensitive probabilistic tracking: lowering
stddev from 0.2 to 0.15.

## Release 1.9

* Tracking on best original peaks or on weighted mean of best original peaks and TOMs (non-public interface).
* **Interface change**: All tracking related commands (whenever you used `--track`) are not part of `TractSeg` anymore
 but now are combined under `Tracking`. Moreover the option `--filter_tracking_by_endpoints` is now activated per
 default. If you want to deactivate is use `--no_filtering_by_endpoints`.
 So the following command

```
TractSeg -i peaks.nii.gz --output_type TOM --track --filter_tracking_by_endpoints
```

becomes

```
TractSeg -i peaks.nii.gz --output_type TOM 
Tracking -i peaks.nii.gz
```

* Works with pytorch 1.0 now
* Bugfixes and minor improvements

## Release 1.8

* "Probabilistic" tracking on TOM output
* New trk format (nibabel.streamlines API). Use `--tracking_format trk` to use it.
* Option to do Mrtrix iFOD2 tracking on original FODs but filter by TractSeg masks.
* Added 3D U-Net, but not used
* Minor improvements & Bugfixes

## Release 1.7.1

* **Interface change**: TractSeg does not automatically flip the peaks anymore if it detects that they probably have
the wrong orientation (The peak check is only correct in 98% of the cases. In the remaining 2% it would incorrectly flip
the peaks and the user would wonder why the results of TractSeg are so bad. Therefore now the user is informed if
TractSeg thinks that flipping is needed, but he has to do it on his own and manually verify the result.) Therefore
the command line option `-deactivate_peak_check` is not needed anymore and removed.
* Bugfixes

## Release 1.7

* Reduced memory consumption of TOM (downside: `--single_output_file` for TOM not working anymore)
* Added more testing
* Number of fibers can be set as parameter
* Brain mask not needed anymore for tracking

tractseg/python_api.py

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import warnings
import importlib
import time
import os
from os.path import join
import numpy as np

from tractseg.libs.system_config import SystemConfig as C
from tractseg.libs.system_config import get_config_name
from tractseg.libs import exp_utils
from tractseg.libs import utils
from tractseg.libs import data_utils
from tractseg.libs import direction_merger
from tractseg.libs import peak_utils
from tractseg.libs import img_utils
from tractseg.data.data_loader_inference import DataLoaderInference
from tractseg.data import dataset_specific_utils
from tractseg.libs import trainer
from tractseg.models.base_model import BaseModel

warnings.simplefilter("ignore", UserWarning)    #hide scipy warnings
warnings.simplefilter("ignore", FutureWarning)    #hide h5py warnings


def run_tractseg(data, output_type="tract_segmentation",
                 single_orientation=False, dropout_sampling=False, threshold=0.5,
                 bundle_specific_postprocessing=True, get_probs=False, peak_threshold=0.1,
                 postprocess=False, peak_regression_part="All", input_type="peaks",
                 blob_size_thr=50, nr_cpus=-1, verbose=False, manual_exp_name=None,
                 inference_batch_size=1, tract_definition="TractQuerier+", bedpostX_input=False,
                 tract_segmentations_path=None, TOM_dilation=1, unit_test=False):
    """
    Run TractSeg

    Args:
        data: input peaks (4D numpy array with shape [x,y,z,9])
        output_type: TractSeg can segment not only bundles, but also the end regions of bundles.
            Moreover it can create Tract Orientation Maps (TOM).
            'tract_segmentation' [DEFAULT]: Segmentation of bundles (72 bundles).
            'endings_segmentation': Segmentation of bundle end regions (72 bundles).
            'TOM': Tract Orientation Maps (20 bundles).
        single_orientation: Do not run model 3 times along x/y/z orientation with subsequent mean fusion.
        dropout_sampling: Create uncertainty map by monte carlo dropout (https://arxiv.org/abs/1506.02142)
        threshold: Threshold for converting probability map to binary map
        bundle_specific_postprocessing: Set threshold to lower and use hole closing for CA nd FX if incomplete
        get_probs: Output raw probability map instead of binary map
        peak_threshold: All peaks shorter than peak_threshold will be set to zero
        postprocess: Simple postprocessing of segmentations: Remove small blobs and fill holes
        peak_regression_part: Only relevant for output type 'TOM'. If set to 'All' (default) it will return all
            72 bundles. If set to 'Part1'-'Part4' it will only run for a subset of the bundles to reduce memory
            load.
        input_type: Always set to "peaks"
        blob_size_thr: If setting postprocess to True, all blobs having a smaller number of voxels than specified in
            this threshold will be removed.
        nr_cpus: Number of CPUs to use. -1 means all available CPUs.
        verbose: Show debugging infos
        manual_exp_name: Name of experiment if do not want to use pretrained model but your own one
        inference_batch_size: batch size (higher: a bit faster but needs more RAM)
        tract_definition: Select which tract definitions to use. 'TractQuerier+' defines tracts mainly by their
            cortical start and end region. 'xtract' defines tracts mainly by ROIs in white matter.
        bedpostX_input: Input peaks are generated by bedpostX
        tract_segmentations_path: path to the bundle_segmentations (only needed for peak regression to remove peaks
            outside of the segmentation mask)
        TOM_dilation: Dilation applied to the tract segmentations before using them to mask the TOMs.

    Returns:
        4D numpy array with the output of tractseg
        for tract_segmentation:     [x, y, z, nr_of_bundles]
        for endings_segmentation:   [x, y, z, 2*nr_of_bundles]
        for TOM:                    [x, y, z, 3*nr_of_bundles]
    """
    start_time = time.time()

    if manual_exp_name is None:
        config = get_config_name(input_type, output_type, dropout_sampling=dropout_sampling,
                                 tract_definition=tract_definition)
        Config = getattr(importlib.import_module("tractseg.experiments.pretrained_models." + config), "Config")()
    else:
        Config = exp_utils.load_config_from_txt(join(C.EXP_PATH,
                                                     exp_utils.get_manual_exp_name_peaks(manual_exp_name, "Part1"),
                                                     "Hyperparameters.txt"))

    # Do not do any postprocessing if returning probabilities (because postprocessing only works on binary)
    if get_probs:
        bundle_specific_postprocessing = False
        postprocess = False

    Config = exp_utils.get_correct_labels_type(Config)
    Config.VERBOSE = verbose
    Config.TRAIN = False
    Config.TEST = False
    Config.SEGMENT = False
    Config.GET_PROBS = get_probs
    Config.LOAD_WEIGHTS = True
    Config.DROPOUT_SAMPLING = dropout_sampling
    Config.THRESHOLD = threshold
    Config.NR_CPUS = nr_cpus
    Config.INPUT_DIM = dataset_specific_utils.get_correct_input_dim(Config)
    Config.RESET_LAST_LAYER = False

    if Config.EXPERIMENT_TYPE == "tract_segmentation" and bundle_specific_postprocessing:
        Config.GET_PROBS = True

    if manual_exp_name is not None and Config.EXPERIMENT_TYPE != "peak_regression":
        Config.WEIGHTS_PATH = exp_utils.get_best_weights_path(join(C.EXP_PATH, manual_exp_name), True)
    else:
        if tract_definition == "TractQuerier+":
            if input_type == "peaks":
                if Config.EXPERIMENT_TYPE == "tract_segmentation" and Config.DROPOUT_SAMPLING:
                    Config.WEIGHTS_PATH = join(C.WEIGHTS_DIR, "pretrained_weights_tract_segmentation_v3.npz")
                elif Config.EXPERIMENT_TYPE == "tract_segmentation":
                    Config.WEIGHTS_PATH = join(C.WEIGHTS_DIR, "pretrained_weights_tract_segmentation_v3.npz")
                elif Config.EXPERIMENT_TYPE == "endings_segmentation":
                    Config.WEIGHTS_PATH = join(C.WEIGHTS_DIR, "pretrained_weights_endings_segmentation_v4.npz")
                elif Config.EXPERIMENT_TYPE == "dm_regression":
                    Config.WEIGHTS_PATH = join(C.WEIGHTS_DIR, "pretrained_weights_dm_regression_v2.npz")
            else:  # T1
                if Config.EXPERIMENT_TYPE == "tract_segmentation":
                    Config.WEIGHTS_PATH = join(C.NETWORK_DRIVE, "hcp_exp_nodes/x_Pretrained_TractSeg_Models",
                                               "TractSeg_T1_125mm_DAugAll", "best_weights_ep142.npz")
                elif Config.EXPERIMENT_TYPE == "endings_segmentation":
                    Config.WEIGHTS_PATH = join(C.WEIGHTS_DIR, "pretrained_weights_endings_segmentation_v1.npz")
                elif Config.EXPERIMENT_TYPE == "peak_regression":
                    Config.WEIGHTS_PATH = join(C.WEIGHTS_DIR, "pretrained_weights_peak_regression_v1.npz")
        else:  # xtract
            if Config.EXPERIMENT_TYPE == "tract_segmentation":
                Config.WEIGHTS_PATH = join(C.WEIGHTS_DIR, "pretrained_weights_tract_segmentation_xtract_v1.npz")
            elif Config.EXPERIMENT_TYPE == "dm_regression":
                Config.WEIGHTS_PATH = join(C.WEIGHTS_DIR, "pretrained_weights_dm_regression_xtract_v1.npz")
            else:
                raise ValueError("bundle_definition xtract not supported in combination with this output type")

    if Config.VERBOSE:
        print("Hyperparameters:")
        exp_utils.print_Configs(Config)

    data = np.nan_to_num(data)

    #runtime on HCP data: 0.9s
    data, seg_None, bbox, original_shape = data_utils.crop_to_nonzero(data)
    # runtime on HCP data: 0.5s
    data, transformation = data_utils.pad_and_scale_img_to_square_img(data, target_size=Config.INPUT_DIM[0],
                                                                      nr_cpus=nr_cpus)

    if Config.EXPERIMENT_TYPE == "tract_segmentation" or Config.EXPERIMENT_TYPE == "endings_segmentation" or \
            Config.EXPERIMENT_TYPE == "dm_regression":
        print("Loading weights from: {}".format(Config.WEIGHTS_PATH))
        Config.NR_OF_CLASSES = len(dataset_specific_utils.get_bundle_names(Config.CLASSES)[1:])
        utils.download_pretrained_weights(experiment_type=Config.EXPERIMENT_TYPE,
                                          dropout_sampling=Config.DROPOUT_SAMPLING,
                                          tract_definition=tract_definition)
        model = BaseModel(Config, inference=True)
        if single_orientation:  # mainly needed for testing because of less RAM requirements
            data_loder_inference = DataLoaderInference(Config, data=data)
            if Config.DROPOUT_SAMPLING or Config.EXPERIMENT_TYPE == "dm_regression" or Config.GET_PROBS:
                seg, _ = trainer.predict_img(Config, model, data_loder_inference, probs=True,
                                                 scale_to_world_shape=False, only_prediction=True,
                                                 batch_size=inference_batch_size, unit_test=unit_test)
            else:
                seg, _ = trainer.predict_img(Config, model, data_loder_inference, probs=False,
                                                 scale_to_world_shape=False, only_prediction=True,
                                                 batch_size=inference_batch_size)
        else:
            seg_xyz, _ = direction_merger.get_seg_single_img_3_directions(Config, model, data=data,
                                                                           scale_to_world_shape=False,
                                                                           only_prediction=True,
                                                                           batch_size=inference_batch_size)
            if Config.DROPOUT_SAMPLING or Config.EXPERIMENT_TYPE == "dm_regression" or Config.GET_PROBS:
                seg = direction_merger.mean_fusion(Config.THRESHOLD, seg_xyz, probs=True)
            else:
                seg = direction_merger.mean_fusion(Config.THRESHOLD, seg_xyz, probs=False)

    elif Config.EXPERIMENT_TYPE == "peak_regression":
        weights = {
            "Part1": "pretrained_weights_peak_regression_part1_v2.npz",
            "Part2": "pretrained_weights_peak_regression_part2_v2.npz",
            "Part3": "pretrained_weights_peak_regression_part3_v2.npz",
            "Part4": "pretrained_weights_peak_regression_part4_v2.npz",
        }
        if peak_regression_part == "All":
            parts = ["Part1", "Part2", "Part3", "Part4"]
            seg_all = np.zeros((data.shape[0], data.shape[1], data.shape[2], Config.NR_OF_CLASSES * 3))
        else:
            parts = [peak_regression_part]
            Config.CLASSES = "All_" + peak_regression_part
            Config.NR_OF_CLASSES = 3 * len(dataset_specific_utils.get_bundle_names(Config.CLASSES)[1:])

        for idx, part in enumerate(parts):
            if manual_exp_name is not None:
                manual_exp_name_peaks = exp_utils.get_manual_exp_name_peaks(manual_exp_name, part)
                Config.WEIGHTS_PATH = exp_utils.get_best_weights_path(
                    join(C.EXP_PATH, manual_exp_name_peaks), True)
            else:
                Config.WEIGHTS_PATH = join(C.WEIGHTS_DIR, weights[part])
            print("Loading weights from: {}".format(Config.WEIGHTS_PATH))
            Config.CLASSES = "All_" + part
            Config.NR_OF_CLASSES = 3 * len(dataset_specific_utils.get_bundle_names(Config.CLASSES)[1:])
            utils.download_pretrained_weights(experiment_type=Config.EXPERIMENT_TYPE,
                                              dropout_sampling=Config.DROPOUT_SAMPLING, part=part,
                                              tract_definition=tract_definition)
            model = BaseModel(Config, inference=True)

            if single_orientation:
                data_loder_inference = DataLoaderInference(Config, data=data)
                seg, _ = trainer.predict_img(Config, model, data_loder_inference, probs=True,
                                                 scale_to_world_shape=False, only_prediction=True,
                                                 batch_size=inference_batch_size)
            else:
                # 3 dir for Peaks -> bad results
                seg_xyz, _ = direction_merger.get_seg_single_img_3_directions(Config, model, data=data,
                                                                                  scale_to_world_shape=False,
                                                                                  only_prediction=True,
                                                                                  batch_size=inference_batch_size)
                seg = direction_merger.mean_fusion_peaks(seg_xyz, nr_cpus=nr_cpus)

            if peak_regression_part == "All":
                seg_all[:, :, :, (idx*Config.NR_OF_CLASSES) : (idx*Config.NR_OF_CLASSES+Config.NR_OF_CLASSES)] = seg

        if peak_regression_part == "All":
            Config.CLASSES = "All"
            Config.NR_OF_CLASSES = 3 * len(dataset_specific_utils.get_bundle_names(Config.CLASSES)[1:])
            seg = seg_all


    if Config.EXPERIMENT_TYPE == "tract_segmentation" and bundle_specific_postprocessing and not dropout_sampling:
        # Runtime ~4s
        seg = img_utils.bundle_specific_postprocessing(seg, dataset_specific_utils.get_bundle_names(Config.CLASSES)[1:])

    # runtime on HCP data: 5.1s
    seg = data_utils.cut_and_scale_img_back_to_original_img(seg, transformation, nr_cpus=nr_cpus)
    # runtime on HCP data: 1.6s
    seg = data_utils.add_original_zero_padding_again(seg, bbox, original_shape, Config.NR_OF_CLASSES)

    if Config.EXPERIMENT_TYPE == "peak_regression":
        seg = peak_utils.mask_and_normalize_peaks(seg, tract_segmentations_path,
                                                  dataset_specific_utils.get_bundle_names(Config.CLASSES)[1:],
                                                  TOM_dilation, nr_cpus=nr_cpus)

    if Config.EXPERIMENT_TYPE == "tract_segmentation" and postprocess and not dropout_sampling:
        # Runtime ~7s for 1.25mm resolution
        # Runtime ~1.5s for  2mm resolution
        st = time.time()
        seg = img_utils.postprocess_segmentations(seg, dataset_specific_utils.get_bundle_names(Config.CLASSES)[1:],
                                                  blob_thr=blob_size_thr, hole_closing=None)

    exp_utils.print_verbose(Config.VERBOSE, "Took {}s".format(round(time.time() - start_time, 2)))
    return seg

